Discussion: turn examples into runnable, testable fragments of code

mroderick opened the pull request:
🚧 do not merge 🚧 

This PR is just an opening for a discussion of how we can turn the API examples into runnable, testable code.

It often happens that when you try to learn to use a new JS package, that the examples our outdated or have typos in them. Sinon is no exception to that.

I'd like to fix this, by making all the examples used in the API documentation for `sinon@5.x` be runnable and have a basic unit test.

The changes should go into `release-source/`, I am just using the `5.0.7` as a working copy, as `release-source/` does not get built by Jekyll.

🚧 do not merge 🚧 

#### How to verify
1. Check out this branch
1. `cd docs`
1. `bundle exec jekyll serve`
1. Navigate to http://localhost:4000/releases/v5.0.7/fakes/
1. Observe the slightly updated examples
1. `npm run test-examples`
1. Observe that the tests pass

fatso83 commented:
I have thought the same thoughts many a times, and doings this is exactly in line with how the [Pragmatic Programmer](https://pragprog.com/book/tpp/the-pragmatic-programmer) was made. They also made sure all the code compiled and built before their docbook conversion :smile: 

Good initiative!

fearphage commented on docs/_includes/examples/.eslintrc.yml:
```diff
@@ -0,0 +1,23 @@
+rules:
+  no-console: off
+  max-nested-callbacks: off
+  no-restricted-syntax: [error, 'TryStatement']
```
`TryStatement` doesn't need quotes.

mroderick commented:
I've updated this PR with runnable, testable examples for pretty much all of the documentation for `fakes`.

I've updated the "How to verify" instructions.

@fatso83 @fearphage it would be great if you could spare the time to review and verify this PR before I continue.

Once there are no more comments for this PR, then I'll update all the `5.x` releases and `release-source` to use these examples and we can merge it.

fatso83 commented on docs/_includes/examples/fakes/__tests.js:
```diff
@@ -0,0 +1,185 @@
+"use strict";
+
+var proxyquire = require("proxyquire").noCallThru();
```
I think there should be a comment here that says why we are using `proxyquire`. Immediately, it didn't make sense to me why we should be stubbing out `sinon` below, until it hit me that it's of course because we want to use the most recent version.

fatso83 commented on docs/_includes/examples/fakes/__tests.js:
```diff
@@ -0,0 +1,185 @@
+    describe("throws", function () {
+        it("should throw Error with message: 'not apple pie'", function () {
+            try {
+                proxyquire("./throws", {
```
I didn't understand why `proxyquire` was necessary at first, ref comment above. Another thing is that I think all these proxyquire calls obscures what is going on. To be honest, if we intend to replace all instances of `sinon` in these tests with a link seam, I think this perfectly validates using the [`@global`](https://github.com/thlorenz/proxyquire#globally-override-require) flag to proxyquire. Using that, the tests would be simpler - most could be one-liners.

Example (this test):

```javascript
assert.exception(require.bind(null, "./throws"), "not apple pie");
```

mroderick replied:
Maybe I am not reading the documentation right, but I am having a hard time seeing how the use of `@global` is going to turn the tests into one liners.

Could try to create a runnable example that demonstrates what you're thinking?

fatso83 replied:
Sorry - my comment doesn't make sense post-morning. Must have been high - on sleep depravation.

fearphage commented on docs/_includes/examples/fakes/__tests.js:
```diff
@@ -0,0 +1,146 @@
+    });
+
+    // basic.js
+    describe("basic invocation", function () {
```
These `describe` blocks here serve no purpose. If you want to make the test names more descriptive, I would suggest you just add to the test names. Could we simplify this a bit? It would be nice to use the names in the `proxyquire` as well although that's not necessary. This could be simplified into an object:

```js
{ 'basic - test name': testFn } 
// or
{ 'test name': ['basic' /* filename to load with proxyquire? */, testFn] }

// ...
Object.keys(testObject)
  .forEach(label => {
    it(label, testFn);
  })
;
```

fearphage replied:
Another thought, why don't we read the names of the files in the test directory and just load and run them dynamically?Then we could use the file names as the key in the `testObject` above:

```js
{  './basic': testFn }
```

mroderick replied:
Good points. I'll try to re-work this first version to incorporate these

fearphage commented:
Please ignore this blip. I'm just closing and reopening this issue to get all the tests to run again. Some may have been updated since the PR was created.

fatso83 commented:
Would love to have this merged (against `release-source`, of course).

mroderick commented:
This was a good discussion, thank you all for participating. I think we've proven that the idea has legs, and so I will add it as a feature for #1651
